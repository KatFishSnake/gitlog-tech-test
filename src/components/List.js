import { useState, useEffect } from "react";
import { Gitgraph, Mode } from "@gitgraph/react";
import { format, isYesterday, isToday } from "date-fns";

import { mockCommitList } from "../mockData";

const getDate = (date) => {
  const timestamp = new Date(date);
  const today = isToday(timestamp);
  const yesterday = isYesterday(timestamp);
  return `${format(
    timestamp,
    `${(today && "Today") || (yesterday && "Yesterday") || "MMM d, yyyy"}`
  )} at ${format(timestamp, "hh:mm a")}`;
};

const List = ({ query }) => {
  const [commitList, setCommitList] = useState([]);

  useEffect(() => {
    console.info(`Filtering by query ${query}`);
    setCommitList(mockCommitList);
  }, [query]);

  return (
    <div className="flex flex-row flex-grow overflow-auto">
      <div className="w-16 flex-col hidden sm:flex flex-none">
        <div className="w-full sticky top-0 px-1 py-3 text-red-900 bg-red-300 flex-0">
          Graph
        </div>
        <div className="bg-red-200 flex-1">graph</div>
      </div>

      <div className="flex flex-col w-full flex-1">
        <div className="sticky top-0 flex-1 flex flex-row">
          <div className=" px-1 py-3 text-red-900 bg-red-300">Description</div>
          <div className=" px-1 py-3 text-red-900 bg-red-300">Commit #</div>
          <div className=" px-1 py-3 text-red-900 bg-red-300">Author</div>
          <div className=" px-1 py-3 text-red-900 bg-red-300">Date</div>
        </div>

        <div className="overflow-auto">
          {commitList.map((item) => (
            <div className="flex flex-row w-full" key={item.hash}>
              <div className="px-1 py-4 truncate">{item.subject}</div>
              <div className="px-1 py-4 truncate">{item.hash}</div>
              <div className="px-1 py-4 truncate">{item.author.name}</div>
              <div className="px-1 py-4 truncate">
                {getDate(item.author.timestamp)}
              </div>
            </div>
          ))}
        </div>
      </div>
    </div>
  );
};

export default List;
