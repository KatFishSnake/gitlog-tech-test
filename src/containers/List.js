import { useMemo } from "react";
import PropTypes from "prop-types";
import { format, isYesterday, isToday } from "date-fns";
import styled from "styled-components";
import clsx from "clsx";

import GitGraph from "../components/GitGraph";
import Loader from "../components/Loader";
import useFetch from "../hooks/useFetch";
import { sizes } from "../style-vars";
import { graphColors, mockCommitList } from "../config";

const gridColumnTemplate =
  "1fr minmax(80px, 101px) minmax(80px, 109px) minmax(100px,194px)";

const StyledRow = styled.div`
  grid-template-columns: 1fr;

  @media ${sizes.md} {
    grid-template-columns: ${gridColumnTemplate};
  }
`;

const StyledHeadRow = styled.div`
  grid-template-columns: 4rem 1fr;

  @media ${sizes.md} {
    grid-template-columns: 4rem ${gridColumnTemplate};
  }
`;

const StyledGitGraph = styled(GitGraph)`
  margin-top: -0.095rem;
`;

const StyledHeaderCell = styled.div`
  padding-top: 5px;
  padding-bottom: 3px;
  border-color: #a9a9a9;
`;

const StyledTag = styled.span`
  ${({ refKey, tagColors }) =>
    tagColors[refKey] ? `background-color: ${tagColors[refKey]};` : ""}
  color: #00000087;
`;

const getTagColors = (list, colors) => {
  let colorCounter = 0;
  const branchToColorMap = {};

  [...list].reverse().forEach((item) => {
    item.refs.forEach((ref) => {
      if (ref.includes("tag: ") || ref.includes("HEAD")) return;
      const withoutOrigin = ref.replace("origin/", "");
      if (!branchToColorMap[withoutOrigin]) colorCounter++;
      branchToColorMap[withoutOrigin] = colors[colorCounter];
      branchToColorMap[`origin/${withoutOrigin}`] = colors[colorCounter];
      if (colorCounter === colors.length) colorCounter = 0;
    });
  });

  return branchToColorMap;
};

const getDate = (date) => {
  const timestamp = new Date(date);
  const today = isToday(timestamp);
  const yesterday = isYesterday(timestamp);
  const formattedMonth =
    (today && "Today") ||
    (yesterday && "Yesterday") ||
    format(timestamp, "MMM d, yyyy");
  return `${formattedMonth} at ${format(timestamp, "hh:mm a")}`;
};

const List = ({ query }) => {
  const queryMemoed = useMemo(() => ({ query }), [query]);
  const { loading, data: commitList } = useFetch(
    mockCommitList,
    queryMemoed,
    1000
  );
  const tagColors = useMemo(() => getTagColors(commitList, graphColors), [
    commitList,
  ]);

  return (
    <div className="flex text-xxs flex-wrap overflow-y-scroll overflow-x-hidden">
      <StyledHeadRow className="sticky top-0 grid grid-cols-4 w-full items-start">
        <StyledHeaderCell className="w-full sticky top-0 px-2 py-1 bg-d-gray flex-0 text-center shadow-inner">
          Graph
        </StyledHeaderCell>
        <StyledHeaderCell className="px-2 py-1 bg-d-gray border-l shadow-inner">
          Description
        </StyledHeaderCell>
        <StyledHeaderCell className="hidden md:block px-2 py-1 bg-d-gray border-l shadow-inner">
          Commit #
        </StyledHeaderCell>
        <StyledHeaderCell className="hidden md:block px-2 py-1 bg-d-gray border-l shadow-inner">
          Author
        </StyledHeaderCell>
        <StyledHeaderCell className="hidden md:block px-2 py-1 bg-d-gray border-l shadow-inner">
          Date
        </StyledHeaderCell>
      </StyledHeadRow>

      {loading ? (
        <Loader className="mt-3 text-center w-full" />
      ) : (
        <>
          <div className="w-16 flex-col flex">
            <div className="flex-1">
              <StyledGitGraph
                graphData={commitList}
                className={clsx(query && "hidden")}
              />
            </div>
          </div>

          <div className="flex flex-col flex-1">
            <div>
              {commitList.map((item) => (
                <StyledRow className="grid grid-cols-4 w-full" key={item.hash}>
                  <div className="px-1 py-1 truncate" title={item.subject}>
                    {item.refs.map((ref) => (
                      <StyledTag
                        refKey={ref}
                        tagColors={tagColors}
                        className="bg-gray-200 rounded-2xl px-2 py-1 mr-1 text-xxxs"
                        key={ref}
                      >
                        {ref}
                      </StyledTag>
                    ))}
                    {item.subject}
                  </div>
                  <div
                    className="px-1 py-1 truncate hidden md:block"
                    title={item.hash}
                  >
                    {item.hash}
                  </div>
                  <div
                    className="px-1 py-1 truncate hidden md:block"
                    title={item.author.name}
                  >
                    {item.author.name}
                  </div>
                  <div className="px-1 py-1 truncate hidden md:block">
                    {getDate(item.author.timestamp)}
                  </div>
                </StyledRow>
              ))}
            </div>
          </div>
        </>
      )}
    </div>
  );
};

List.propTypes = {
  query: PropTypes.string,
};

export default List;
