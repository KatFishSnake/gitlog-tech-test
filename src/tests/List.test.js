import { render } from "@testing-library/react";

import List from "../containers/List";

test("Renders List consistently", () => {
  const { container } = render(<List />);
  expect(container).toMatchSnapshot();
});
