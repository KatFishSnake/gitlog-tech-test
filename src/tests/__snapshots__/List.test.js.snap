// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders List consistently 1`] = `
<div>
  <div
    class="flex text-xxs flex-wrap overflow-y-scroll overflow-x-hidden"
  >
    <div
      class="sc-dlfnbm eZAuYd sticky top-0 grid grid-cols-4 w-full items-start"
    >
      <div
        class="sc-eCssSg eVfvmO w-full sticky top-0 px-2 py-1 bg-d-gray flex-0 text-center shadow-inner"
      >
        Graph
      </div>
      <div
        class="sc-eCssSg eVfvmO px-2 py-1 bg-d-gray border-l shadow-inner"
      >
        Description
      </div>
      <div
        class="sc-eCssSg eVfvmO hidden md:block px-2 py-1 bg-d-gray border-l shadow-inner"
      >
        Commit #
      </div>
      <div
        class="sc-eCssSg eVfvmO hidden md:block px-2 py-1 bg-d-gray border-l shadow-inner"
      >
        Author
      </div>
      <div
        class="sc-eCssSg eVfvmO hidden md:block px-2 py-1 bg-d-gray border-l shadow-inner"
      >
        Date
      </div>
    </div>
    <span
      class="mt-3 text-center w-full"
    >
      Loading...
    </span>
  </div>
</div>
`;
